File: src/services/parser/v2/text.ts
import { StringDate } from "../../../utils";

const DAY_NAMES = [
    'понедельник',
    'вторник',
    'среда',
    'четверг',
    'пятница',
    'суббота',
    'воскресенье'
];

const DATE_RE = /(\d{1,2}\.\d{1,2}\.\d{2,4})/;

export function cleanText(text?: string | null): string | null {
    if (text == null) {
        return null;
    }

    const normalized = text.replace(/\s+/g, ' ').trim();
    return normalized.length ? normalized : null;
}

export function normalizeDate(value: string): string | null {
    const parts = value.split('.').map((part) => part.trim()).filter(Boolean);
    if (parts.length < 3) {
        return null;
    }

    let [day, month, year] = parts;
    if (year.length === 2) {
        year = `20${year}`;
    }

    const dayNum = Number(day);
    const monthNum = Number(month);
    const yearNum = Number(year);

    if (!dayNum || !monthNum || !yearNum) {
        return null;
    }

    const normalized = `${dayNum.toString().padStart(2, '0')}.${monthNum.toString().padStart(2, '0')}.${yearNum}`;
    try {
        return StringDate.fromStringDate(normalized).toStringDate();
    } catch {
        return null;
    }
}

export function parseDayLabel(value?: string | null): { day: string, weekday?: string } | null {
    const text = cleanText(value);
    if (!text) {
        return null;
    }

    const dateMatch = text.match(DATE_RE);
    if (!dateMatch) {
        return null;
    }

    const normalized = normalizeDate(dateMatch[1]);
    if (!normalized) {
        return null;
    }

    const lower = text.toLowerCase();
    const weekdayMatch = DAY_NAMES.find((name) => lower.includes(name));
    const weekday = weekdayMatch ? capitalize(weekdayMatch) : undefined;

    return {
        day: normalized,
        weekday
    };
}

function capitalize(text: string): string {
    return text.charAt(0).toUpperCase() + text.slice(1);
}

export function extractLines(cell?: HTMLTableCellElement | null): string[] {
    if (!cell) {
        return [];
    }

    const lines: string[] = [];
    let buffer = '';

    const flush = () => {
        const normalized = cleanText(buffer);
        if (normalized) {
            lines.push(normalized);
        }
        buffer = '';
    };

    const walk = (node: Node) => {
        if (node.nodeType === node.TEXT_NODE) {
            buffer += node.textContent ?? '';
            return;
        }

        if (node.nodeType !== node.ELEMENT_NODE) {
            return;
        }

        const element = node as Element;
        const tag = element.tagName.toLowerCase();

        if (tag === 'br') {
            flush();
            return;
        }

        const isBlock = tag === 'p' || tag === 'div' || tag === 'tr' || tag === 'li';

        if (isBlock) {
            flush();
        }

        for (const child of Array.from(element.childNodes)) {
            walk(child);
        }

        if (isBlock) {
            flush();
        }
    };

    walk(cell);
    flush();

    return lines;
}

export function parseLessonNumber(value?: string | null): number | null {
    const text = cleanText(value);
    if (!text) {
        return null;
    }

    const match = text.match(/(\d+)/);
    if (!match) {
        return null;
    }

    const num = Number(match[1]);
    return Number.isFinite(num) ? num : null;
}

export function isTeacherLine(value: string): boolean {
    const text = value.trim();
    if (!text) {
        return false;
    }

    return /^[A-ZА-ЯЁ][a-zа-яё-]+\s+[A-ZА-ЯЁ]\.\s*[A-ZА-ЯЁ]?\.?$/.test(text);
}

File: internal/parser/v2/text.go
package v2

import (
	"regexp"
	"strconv"
	"strings"
	"time"
	"unicode"

	"github.com/PuerkitoBio/goquery"
)

var dateRe = regexp.MustCompile(`(\d{1,2}\.\d{1,2}\.\d{2,4})`)
var dayNames = []string{
	"понедельник", "вторник", "среда", "четверг",
	"пятница", "суббота", "воскресенье",
}

func cleanText(text string) string {
	if text == "" {
		return ""
	}
	normalized := strings.TrimSpace(text)
	normalized = strings.Join(strings.Fields(normalized), " ")
	if len(normalized) == 0 {
		return ""
	}
	return normalized
}

func normalizeDate(value string) string {
	parts := strings.Split(value, ".")
	if len(parts) < 3 {
		return ""
	}
	day, _ := strconv.Atoi(strings.TrimSpace(parts[0]))
	month, _ := strconv.Atoi(strings.TrimSpace(parts[1]))
	yearStr := strings.TrimSpace(parts[2])
	if len(yearStr) == 2 {
		yearStr = "20" + yearStr
	}
	year, _ := strconv.Atoi(yearStr)
	if day == 0 || month == 0 || year == 0 {
		return ""
	}
	t := time.Date(year, time.Month(month), day, 0, 0, 0, 0, time.UTC)
	return t.Format("02.01.2006")
}

func parseDayLabel(value string) (day string, weekday string) {
	text := cleanText(value)
	if text == "" {
		return "", ""
	}
	dateMatch := dateRe.FindString(text)
	if dateMatch == "" {
		return "", ""
	}
	normalized := normalizeDate(dateMatch)
	if normalized == "" {
		return "", ""
	}
	lower := strings.ToLower(text)
	for _, name := range dayNames {
		if strings.Contains(lower, name) {
			weekday = capitalize(name)
			break
		}
	}
	return normalized, weekday
}

func capitalize(s string) string {
	if s == "" {
		return s
	}
	runes := []rune(s)
	return string([]rune{unicode.ToUpper(runes[0])}) + string(runes[1:])
}

func extractLines(cell *goquery.Selection) []string {
	if cell == nil || cell.Length() == 0 {
		return nil
	}

	var lines []string
	var buffer strings.Builder

	flush := func() {
		normalized := cleanText(buffer.String())
		if normalized != "" {
			lines = append(lines, normalized)
		}
		buffer.Reset()
	}

	cell.Contents().Each(func(_ int, n *goquery.Selection) {
		if goquery.NodeName(n) == "br" {
			flush()
			return
		}
		text := n.Text()
		buffer.WriteString(text)
	})
	flush()

	return lines
}

func parseLessonNumber(value string) int {
	text := cleanText(value)
	if text == "" {
		return 0
	}
	re := regexp.MustCompile(`(\d+)`)
	match := re.FindString(text)
	if match == "" {
		return 0
	}
	num, _ := strconv.Atoi(match)
	return num
}

func isTeacherLine(value string) bool {
	text := strings.TrimSpace(value)
	if text == "" {
		return false
	}
	re := regexp.MustCompile(`^[А-ЯЁA-Z][а-яёa-z\-]+\s+[А-ЯЁA-Z]\.\s*[А-ЯЁA-Z]?\.?$`)
	return re.MatchString(text)
}
